families/track_and_trade/sawtooth_track_and_trade/handler.py

```diff
@@ -0,0 +1,633 @@
+    LOGGER.debug(
+        'Creating agent -- key: {} -- name: {}'.format(
+            signer[:6], name))
+
+    agent = Agent(
+        public_key=signer,
+        name=name,
+        timestamp=timestamp,
+    )
+
+    container.entries.extend([agent])
+    container.entries.sort(key=lambda ag: ag.public_key)
+
+    _set_container(state, address, container)
+
+
+def _create_record(payload, signer, timestamp, state):
+    '''
+    * Check that the signer is registered.
+    * Check that the record doesn't already exist.
+    * Check that the record type exists.
+    * Check that the required properties have been provided.
+    * Create the record (and record container if needed).
+    * Create the associated properties.
+    * Post the provided property values.
+    '''
+
+    # Check that the signer is registered.
+    _verify_agent(state, signer)
+
+    txn = payload.create_record
+
+    # Check that the record doesn't already exist
+    record_id = txn.record_id
+
+    record_address = Addressing.make_record_address(record_id)
+    record_container = _get_container(state, record_address)
+
+    if record_container is not None:
+        if any([record.record_id == record_id
+                for record in record_container.entries]):
+            raise InvalidTransaction(
+                'Record {} already exists'.format(record_id))
+
+    # Check that the record type exists.
+    record_type = txn.record_type
+    type_address = Addressing.make_record_type_address(record_type)
+    type_container = _get_container(state, type_address)
+
+    if type_container is None or all([
+            rec_type.name != record_type
+            for rec_type in type_container.entries]):
+        raise InvalidTransaction(
+            'No record type {} exists'.format(record_type))
+
+    # Record type exists -- check the required properties
+    for rec_type in type_container.entries:
+        if rec_type.name == record_type:
+            type_properties = rec_type.properties
+            break
+
+    type_schemata = {
+        prop.name: prop
+        for prop in type_properties
+    }
+
+    required_properties = [
+        prop
+        for prop in type_properties
+        if prop.required
+    ]
+
+    provided_properties = {
+        prop.name: prop
+        for prop in txn.properties
+    }
+
+    # Make sure the required properties are all provided
+    required_names = [prop.name for prop in required_properties]
+
+    for name in required_names:
+        if name not in provided_properties:
+            raise InvalidTransaction(
+                'Required property {} not provided'.format(
+                    name))
+
+    # Make sure the provided properties have the right type
+    required_types = {
+        prop.name: prop.data_type
+        for prop in required_properties
+    }
+
+    provided_types = {
+        name: prop.data_type
+        for name, prop in provided_properties.items()
+    }
+
+    for provided in provided_properties:
```

Comment:
The name `provided` implies to me you are iterating through the properties, not their names.

Reply:
Also, couldn't this be done in the same loop as making sure they exist? Could simplify things.

Reply:
The two checks loop over different sets -- one loops over the required properties and makes sure they have all been provided, the other loops over the provided properties and makes sure they have the correct type.